Update SerialPort.Linux.cs

Johonis opened the pull request:
add ttyGSX devices to supported devices to use also generic serial devices

Commit:
Update SerialPort.Linux.cs

add ttyGSX devices to supported devices

stephentoub commented on src/libraries/System.IO.Ports/src/System/IO/Ports/SerialPort.Linux.cs:
```diff
@@ -33,11 +33,13 @@ public static string[] GetPortNames()
                         (!isTtyS && Directory.Exists(entry.FullName + "/device/tty")) ||
-                        Directory.Exists(sysUsbDir + entry.Name))
+                        Directory.Exists(sysUsbDir + entry.Name) ||
+                        isTtyGS)
```
For `isTtys`, there are several additional checks being done when it's true.  No similar checks are needed when `isTtyGS` is true?

Johonis replied:
As described in issue #1677, I can only say that the port structure does not necessarily behave like ttySx. As mentioned by @krwq, GigaSet phones also use this serial port and I have no idea what features their serial device has.
We could check if the folder "/sys/class/tty/ttyGS0/dev" exists. However, the fact that this folder exists on my system does not mean that it exists on other ports.
So I would suggest we keep this check.

krwq replied:
I generally prefer being more restrictive with checks and fix it when someone reports the problem but at the same time `ttyGS` seem to be documented to be related with serial ports. `ttyS` have a lot of stuff which you normally don't want to show up when enumerating ports so I don't really feel strongly either way

stephentoub commented on src/libraries/System.IO.Ports/src/System/IO/Ports/SerialPort.Linux.cs:
```diff
@@ -33,11 +33,13 @@ public static string[] GetPortNames()
                     // checking for that for non-ttyS entries is incorrect as some uart
                     // devices are incorrectly filtered out
                     bool isTtyS = entry.Name.StartsWith("ttyS");
+                    bool isTtyGS = entry.Name.StartsWith("ttyGS");
```
isTtyGS can't be true if isTtyS is true, as they're checking for competing conditions.
```suggestion
                    bool isTtyGS = !isTtyS && entry.Name.StartsWith("ttyGS");
```

Johonis committed:
Apply suggestions from code review

Entry can only be ttyGSx. when its not ttySx

Co-Authored-By: Stephen Toub <stoub@microsoft.com>

Commit:
Update SerialPort.Linux.cs

ttyGSx creates a file in "/sys/class/tty/ttyGS0" named "dev" - use this file as an additional check

Johonis commented:
I just added an additional check, like @krwq mentioned in the last conversation.

wfurt reviewed (approved):
LGTM.

lpereira commented on src/libraries/System.IO.Ports/src/System/IO/Ports/SerialPort.Linux.cs:
```diff
@@ -33,11 +33,13 @@ public static string[] GetPortNames()
                     // checking for that for non-ttyS entries is incorrect as some uart
                     // devices are incorrectly filtered out
                     bool isTtyS = entry.Name.StartsWith("ttyS");
+                    bool isTtyGS = !isTtyS && entry.Name.StartsWith("ttyGS");
```
Might be better to have a helper function to match not only this one (and also use in the path where `/dev` is enumerated) against a regex.  For instance, `tty[A-Za-z]+[0-9]+` should match all non-terminal TTY devices (which would match `tty[0-9]+`).  We'll never be able to keep up with new device names.

krwq replied:
I'd recommend creating an issue about it with suggested solution. So far we haven't been able to find any way to properly detect it other than combination of name and probing for some files. Also I don't believe new device types show up that often so I don't find it as super high priority. Note that this is a convenience API and if device is in fact TTY then passing it to constructor will work even if it's not listed by GetPortNames

Johonis commented:
I am not yet completely familiar with pull requests on github. When it's going to be merged?

carlossanlop commented:
This PR fixed: https://github.com/dotnet/runtime/issues/1677
